server/methods/messageSearch.js

```diff
@@ -182,17 +182,25 @@ Meteor.methods({
 			query._hidden = {
 				$ne: true // don't return _hidden messages
 			};
-			if (rid != null) {
+			if (rid) {
 				query.rid = rid;
+				// check if user can access rid room
 				if (Meteor.call('canAccessRoom', rid, currentUserId) !== false) {
-					if (!RocketChat.settings.get('Message_ShowEditedStatus')) {
-						options.fields = {
-							'editedAt': 0
-						};
-					}
-					result.messages = RocketChat.models.Messages.find(query, options).fetch();
+					return result;
 				}
+			} else {
+				query.rid = {
+					$in : RocketChat.models.Rooms.findByContainingUsername(currentUserName)
```

Comment:
This rely's on the username on the room object.  I'm not sure if we plan to keep here. 

I do worry about how heavy this could be...

Our community server for example we have 145k users.  I can only imagine this plus then querying across all rooms in... could be very heavy.

Reply:
How about if global search would be admin-enableable feature? 
I see that on large installations this needs highly more optimized search function but on smaller environments current implementation could be acceptable.

Reply:
@RocketChat/core what do you guys think?  Should we have this as something that can be turned on?

Reply:
I'm not in a position to test this feature, nor do I understand what this bit of code does. Is the proposed new search always doing a global search? Perhaps defaulting to searching the current channel, and requiring the user to select a global search option would result in very few global searches being done, ie only when necessary.

Reply:
Well, in large environments global search could be used as DoS-attack by normal users.
Therefore allowing user to enable global search while searching should be only secondary, admin should decide this primarily.

On smaller environments (depends on multiple factors I think) your point still stands.

If time permits, I'll add option to admin.

Reply:
That works fine for us, our environment is small, and we will enable the option. I would hate to be in a big environment and denied access to global search just because it might be abused.

Compromises could include allowing users to specify up to, say, 5 channels to include in their search, or a search string syntax that can specify up to 5 channels to search. Eg "channel:general,random".

Or, as I suggested earlier, an option to repeat the current search on the next channel in the list.